client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.spooling.encoding;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.google.common.base.VerifyException;
+import com.google.common.io.Closer;
+import io.trino.client.spooling.encoding.JsonDecodingUtils.TypeDecoder;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.fasterxml.jackson.core.JsonParser.Feature.AUTO_CLOSE_SOURCE;
+import static com.fasterxml.jackson.core.JsonParser.Feature.USE_FAST_BIG_NUMBER_PARSER;
+import static com.fasterxml.jackson.core.JsonParser.Feature.USE_FAST_DOUBLE_PARSER;
+import static com.fasterxml.jackson.core.JsonToken.END_ARRAY;
+import static com.fasterxml.jackson.core.JsonToken.START_ARRAY;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterators.unmodifiableIterator;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+
+public class JsonQueryDataAccess
+        implements QueryDataAccess
+{
+    private static final JsonFactory JSON_FACTORY = createJsonFactory();
+
+    private final InputStream stream;
+    private final TypeDecoder[] decoders;
+
+    public JsonQueryDataAccess(TypeDecoder[] decoders, InputStream stream)
+    {
+        this.decoders = requireNonNull(decoders, "decoders is null");
+        this.stream = requireNonNull(stream, "stream is null");
+    }
+
+    @Override
+    public Iterable<List<Object>> toIterable()
+            throws IOException
+    {
+        return new RowWiseIterator(stream, decoders);
+    }
+
+    private static class RowWiseIterator
+            implements Iterable<List<Object>>, Iterator<List<Object>>
+    {
+        private final Closer closer = Closer.create();
+        private boolean closed;
+        private final JsonParser parser;
+        private final TypeDecoder[] decoders;
+
+        public RowWiseIterator(InputStream stream, TypeDecoder[] decoders)
+                throws IOException
+        {
+            this.parser = JSON_FACTORY.createParser(stream);
+            this.decoders = requireNonNull(decoders, "decoders is null");
+            closer.register(stream);
+            closer.register(parser);
+            try {
+                verify(parser.nextToken() == START_ARRAY, "Expected start of array, but got %s", parser.currentToken());
+                verify(parser.nextToken() == START_ARRAY, "Expected start of array, but got %s", parser.currentToken());
```

Comment:
duplicate line

Reply:
Oh - that is intntiuonal :)
Please add some comment and maybe make error messages distinguishable

Reply:
Actually now, we want to test that JSON starts with `[[`